bpf/bpf_host.c

```diff
@@ -373,14 +369,6 @@ handle_ipv6_cont(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
 		return DROP_UNROUTABLE;
 	}
 
-#ifdef ENABLE_IPSEC
-	if (info->key && info->tunnel_endpoint) {
-		__u8 key = get_min_encrypt_key(info->key);
-
-		set_encrypt_key_meta(ctx, key, info->node_id);
-		set_identity_meta(ctx, secctx);
-	}
-#endif
```

Comment:
Could you provide a bit more context on this - it's not constrained to `TUNNEL_MODE`, so what traffic did we even expect here? :thinking:

Reply:
Hm :thinking: 

This code is only executed if going through cilium_host and coming from the host (see `!from_host` check above). AFAIK, for remote destinations, we only go through cilium_host if the destination is part of a remote pod CIDR and we are running in tunneling mode. In native routing mode, we go straight to the native device.

Example routing table for tunneling (10.0.0.0/24 is the remote pod CIDR):
```
10.0.0.0/24 via 10.0.1.61 dev cilium_host src 10.0.1.61 mtu 1373 <- we follow this
10.0.1.0/24 via 10.0.1.61 dev cilium_host src 10.0.1.61
10.0.1.61 dev cilium_host scope link 
192.168.56.0/24 dev enp0s8 proto kernel scope link src 192.168.56.11**
```
Example routing table for native routing:
```
10.0.0.0/24 via 192.168.56.12 dev enp0s8 <- we follow this
10.0.1.0/24 via 10.0.1.61 dev cilium_host src 10.0.1.61 
10.0.1.61 dev cilium_host scope link 
192.168.56.0/24 dev enp0s8 proto kernel scope link src 192.168.56.11
```

So, as far as I can see, this code is only used for tunneling with IPsec. But obviously, tunneling mode should already be handled by the other bit of code I removed just above.

So why was this added? It was added in commit https://github.com/cilium/cilium/commit/b76e6eb59ffcb6af3cd17275762d859450f9a7aa and that commit explicitly says it is to support "direct routing modes". I found that very suspicious because, per the above, in native routing mode, traffic from the hostns shouldn't even go through cilium_host.

I thus tested it out. I've checked IPsec with native routing mode, with and without endpoint routes. I can confirm that, in all cases, traffic from the hostns is not encrypted when going to a remote pod.

AFAICT, this code is dead. I'm unsure when it died.

Reply:
:amaze: thanks for digging deeper!

If you need to push once more - could you split this change (with its related cleanups) into a separate patch? I don't mind cleaning up dead & confusing code as part of the PR, but it would be good to emphasize that it's not contributing to the actual bugfix for Gray's issue. And makes a potential revert down the road a lot easier.